include/nbl/builtin/hlsl/emulated/float64_t_impl.hlsl

```diff
@@ -0,0 +1,312 @@
+#ifndef __HLSL_VERSION
+    return nbl::hlsl::findMSB(val);
+#else
+    return firstbithigh(val);
+#endif
+}
+
+template <>
+inline int _findMSB(uint64_t val)
+{
+#ifndef __HLSL_VERSION
+    return nbl::hlsl::findMSB(val);
+#else
+    int msbHigh = firstbithigh(uint32_t(val >> 32));
+    int msbLow = firstbithigh(uint32_t(val));
+    return msbHigh != -1 ? msbHigh + 32 : msbLow;
+#endif
+}
+
+inline uint64_t castToUint64WithFloat64BitPattern(uint64_t val)
+{
+    if (isZero(val))
+        return val;
+
+    int exp = _findMSB(val);
+    uint64_t mantissa;
+
+    int shiftCnt = 52 - exp;
+    if (shiftCnt >= 0)
+    {
+        mantissa = val << shiftCnt;
+    }
+    else
+    {
+        const int shiftCntAbs = -shiftCnt;
+        uint64_t roundingBit = 1ull << (shiftCnt - 1);
+        uint64_t stickyBitMask = roundingBit - 1;
+        uint64_t stickyBit = val & stickyBitMask;
+
+        mantissa = val >> shiftCntAbs;
+
+        if ((val & roundingBit) && (!stickyBit))
+        {
+            bool isEven = mantissa & 1;
+            if (!isEven)
+                mantissa++;
+        }
+        else if ((val & roundingBit) && (stickyBit || (mantissa & 1)))
+            val += roundingBit;
+
+        //val += (1ull << (shiftCnt)) - 1;
+        //mantissa = val >> shiftCntAbs;
+
+        if (mantissa & 1ull << 53)
+        {
+            mantissa >>= 1;
+            exp++;
+        }
+    }
+    mantissa &= ieee754::traits<float64_t>::mantissaMask;
+    const uint64_t biasedExp = uint64_t(ieee754::traits<float64_t>::exponentBias + exp) << ieee754::traits<float64_t>::mantissaBitCnt;
+
+    return biasedExp | mantissa;
+};
+
+inline uint64_t castToUint64WithFloat64BitPattern(int64_t val)
+{
+    const uint64_t sign = val & ieee754::traits<float64_t>::signMask;
+    const uint64_t absVal = uint64_t(abs(val));
+    return sign | castToUint64WithFloat64BitPattern(absVal);
+};
+
+NBL_CONSTEXPR_INLINE_FUNC uint32_t2 umulExtended(uint32_t lhs, uint32_t rhs)
+{
+    uint64_t product = uint64_t(lhs) * uint64_t(rhs);
+    uint32_t2 output;
+    output.x = uint32_t((product & 0xFFFFFFFF00000000) >> 32);
+    output.y = uint32_t(product & 0x00000000FFFFFFFFull);
+    return output;
+}
+
+NBL_CONSTEXPR_INLINE_FUNC uint64_t flushDenormToZero(uint64_t extractedBiasedExponent, uint64_t value)
+{
+    return extractedBiasedExponent ? value : ieee754::extractSignPreserveBitPattern(value);
+}
+
+NBL_CONSTEXPR_INLINE_FUNC uint64_t assembleFloat64(uint64_t signShifted, uint64_t expShifted, uint64_t mantissa)
+{
+    return  signShifted | expShifted | mantissa;
+}
+
+NBL_CONSTEXPR_INLINE_FUNC bool areBothInfinity(uint64_t lhs, uint64_t rhs)
+{
+    lhs &= ~ieee754::traits<float64_t>::signMask;
+    rhs &= ~ieee754::traits<float64_t>::signMask;
+
+    return lhs == rhs && lhs == ieee754::traits<float64_t>::inf;
+}
+
+NBL_CONSTEXPR_INLINE_FUNC bool areBothZero(uint64_t lhs, uint64_t rhs)
+{
+    return ((lhs << 1) == 0ull) && ((rhs << 1) == 0ull);
```

Comment:
why not `bool((lhs|rhs)&~signMask)` or `bool((lhs|rhs)<<1)`

Reply:
done

Reply:
where?